src/runtime-rs/crates/resource/src/volume/block_volume.rs

```diff
@@ -4,34 +4,146 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use anyhow::Result;
+use std::{collections::HashMap, fs, path::Path, sync::Arc};
 
-use super::Volume;
+use crate::share_fs::{do_get_guest_path, do_get_host_path};
 
-pub(crate) struct BlockVolume {}
+use super::{share_fs_volume::generate_mount_path, Volume};
+use agent::Storage;
+use anyhow::{anyhow, Context, Result};
+use hypervisor::{
+    device_manager::{
+        DeviceManager, KATA_BLK_DEV_TYPE, KATA_MMIO_BLK_DEV_TYPE, VIRTIO_BLOCK, VIRTIO_MMIO,
+    },
+    GenericConfig, Hypervisor,
+};
+use nix::sys::stat;
+use tokio::sync::RwLock;
+pub(crate) struct BlockVolume {
+    storage: Option<agent::Storage>,
+    mount: oci::Mount,
+    device_id: String,
+}
 
 /// BlockVolume: block device volume
 impl BlockVolume {
-    pub(crate) fn new(_m: &oci::Mount) -> Result<Self> {
-        Ok(Self {})
+    pub(crate) async fn new(
+        d: Arc<RwLock<DeviceManager>>,
+        h: &dyn Hypervisor,
+        m: &oci::Mount,
+        read_only: bool,
+        cid: &str,
+        sid: &str,
+    ) -> Result<Self> {
+        let fstat = stat::stat(m.source.as_str()).context(format!("stat {}", m.source))?;
+        info!(sl!(), "device stat: {:?}", fstat);
+        let mut options = HashMap::new();
+        if read_only {
+            options.insert("read_only".to_string(), "true".to_string());
+        }
+        let device_id = d
+            .write()
+            .await
+            .try_add_device(
+                &mut GenericConfig {
+                    host_path: m.source.clone(),
+                    container_path: m.destination.clone(),
+                    dev_type: "b".to_string(),
+                    major: stat::major(fstat.st_rdev) as i64,
+                    minor: stat::minor(fstat.st_rdev) as i64,
+                    file_mode: 0,
+                    uid: 0,
+                    gid: 0,
+                    id: "".to_string(),
+                    bdf: None,
+                    driver_options: options,
+                    io_limits: None,
+                    ..Default::default()
```

Comment:
Some of these specified values are already included in `..Default::default()` so I think this (and other instances) could be simplified?

Reply:
Some variables in the struct is not specified here, this pr just support the basic functionality, so we still need them.